### PR TITLE
TSW 71902 gracefully fail on vCPU quota limits

### PR DESCRIPTION
**Pull request by rohml**
There are 8 cases when going through this code that are captured:
3682 - existing deploy, same region [3 cores]
3099 - existing deploy, new region [3 cores]
3814 - new deploy, existing domain, existing rg, entered # [3 cores]
	   new deploy, new domain, existing rg, entered # [6 cores]
3824 - new deploy, existing domain, existing rg, entered name [3 cores]
	   new deploy, new domain, existing rg, entered name [6 cores]
3851 - new deploy, new domain, new rg [6 cores]
           new deploy, existing domain, new rg [3 cores]	

The locations are placed to preempt the creation of empty resource groups in Azure.

### CODE REVIEW

File: Deploy-CAM.ps1 [hunk 1]
```diff
@@ -3070,6 +3092,13 @@ function Set-VnetConfig() {
 
     # Now select subnets
     $vnet = Get-AzureRmVirtualNetwork -Name $vnetName -ResourceGroupName $vnetRgName
+    
+    # Only check if we are deploying to an existing CAM deployment, not to an existing domain controller.
+    # Domain controller version can use a different location, so we check that one later.
+    if(-not $deployOverDC){
+        Check-Location-Cores -rgLocation $vnet.Location -neededCores $minCoresAddConn
```
Comment (bill-dall) [inline Deploy-CAM.ps1]: How is $minCoresAddConn set here?

**Comment by bill-dall** [comment]
Looks reasonable with a minor question. I'll let others do a fine-tuned review. Thanks for double checking the core counts with me.

File: Deploy-CAM.ps1 (hunk 1 above)
Comment (rohml) [reply Deploy-CAM.ps1]: $minCoresAddConn and $minCoresFullDeploy are both just hardcoded at 3 and 6 respectively. They will need to be updated manually with any changes to the default VMs.

**Comment by rohml** [comment]
Canadaeast is currently fully occupied with offline VMs for testing purposes.

File: Deploy-CAM.ps1 [hunk 2]
```diff
@@ -3012,6 +3012,28 @@ function Write-Host-Warning() {
...
+    # "cores" is the Name.Value for "Total Regional vCPUs"
+    $totalRegionUsage = Get-AzureRmVMUsage -Location $rgLocation | Where-Object {$_.Name.Value -eq "cores"}
+    $availableCores = $totalRegionUsage.Limit - $totalRegionUsage.CurrentValue
+    if($availableCores -lt $neededCores){
+        if($availableCores -eq 1){
+            $areis = "is"
+        }else{
+            $areis = "are" 
+        }
+        Write-Host-Warning "Not enough vCPUs available in $rgLocation. $neededCores vCPUs are required, $availableCores $areis available."
```
Comment (cptera) [inline Deploy-CAM.ps1]: You could probably get rid of the areis if you rephrased this to something like "_ vCPUs are required (_ available)" or soemthing

File: Deploy-CAM.ps1 [hunk 3]
```diff
@@ -3455,6 +3484,9 @@ $vnetConfig.CSsubnetName = $ConnectionServiceSubnetName
 $vnetConfig.GWsubnetName = $GatewaySubnetName
 $vnetConfig.RWsubnetName = $RemoteWorkstationSubnetName
 
+# Set the minimum vCPUs needed for a full deployment or an add-on connection
+$minCoresFullDeploy = 6
```
Comment (cptera) [inline Deploy-CAM.ps1]: Maybe define these at the top of the file instead of hidden in the middle of the file?

Comment (rohml) [reply Deploy-CAM.ps1]: Checked with Bill and added a global variables section to the top.

File: Deploy-CAM.ps1 (hunk 2 above)
Comment (rohml) [reply Deploy-CAM.ps1]: Fixed, now reeds "... _ vCPUs are required, _ available."

File: Deploy-CAM.ps1 (hunk 1 above)
Comment (rohml) [reply Deploy-CAM.ps1]: Updated to pass through parameters to the Set-VnetConfig function.

**Pull request merged**